hibernate-jcache/src/test/java/org/hibernate/orm/test/jcache/QueryCacheWithFilterTest.java

```diff
@@ -0,0 +1,143 @@
+package org.hibernate.orm.test.jcache;
+
+import java.util.List;
+
+import org.hibernate.annotations.Filter;
+import org.hibernate.annotations.FilterDef;
+import org.hibernate.annotations.ParamDef;
+import org.hibernate.cache.spi.RegionFactory;
+import org.hibernate.cfg.AvailableSettings;
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.query.Query;
+import org.hibernate.stat.CacheRegionStatistics;
+import org.hibernate.stat.spi.StatisticsImplementor;
+
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.ServiceRegistry;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.hibernate.testing.orm.junit.Setting;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DomainModel(annotatedClasses = {
+		QueryCacheWithFilterTest.Person.class,
+})
+@SessionFactory(generateStatistics = true)
+@ServiceRegistry(
+		settings = {
+				@Setting(name = AvailableSettings.USE_QUERY_CACHE, value = "true"),
+				@Setting(name = AvailableSettings.USE_SECOND_LEVEL_CACHE, value = "true"),
+				@Setting(name = AvailableSettings.CACHE_REGION_FACTORY, value = "jcache")
+		}
+)
+public class QueryCacheWithFilterTest {
+
+	@BeforeEach
+	public void setUp(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					Person p = new Person();
+					p.setName( "John" );
+					session.persist( p );
+				}
+		);
+	}
+
+	@Test
+	public void testCachedQueryWithFilter(SessionFactoryScope scope) {
+		executeQueryWithCache( scope, false, 0 );
+		executeQueryWithCache( scope, false, 1 );
+		evictQueryRegion(scope);
+		executeQueryWithCache( scope, true, 0 );
+		executeQueryWithCache( scope, true, 1 );
```

Comment:
To actually provoke the bug you'd need to mutate the enabled filters of a session after a query result was cached. Let's try this:

```java
scope.inTransaction(
	session -> {
		Query<Person> query = session.createQuery(
			"from Person p",
			Person.class
		);
		query.setCacheable( true );

		List<Person> resultList = query.getResultList();
		assertThat( resultList ).hasSize( 1 );

		session.enableFilter( "personName" ).setParameter( "name", "John" );
	}
);
scope.inTransaction(
	session -> {
		Query<Person> query = session.createQuery(
			"from Person p",
			Person.class
		);
		query.setCacheable( true );

		List<Person> resultList = query.getResultList();
		assertThat( resultList ).hasSize( 1 );

		assertThat( getQueryCacheRegionStatistics( session ).getHitCount() ).isEqualTo( 1 );
	}
);
```

Reply:
Sorry I'm lost here:
- What I'm trying to test is that QueryKey is properly serializable so it can be used as a cache key => if it's not then the query is not cached, there's a miss the second time, my test validates that
- I don't understand what your test is trying to validate: how is mutating the filter relevant in this case?

Actually, I might have found a second bug: I noticed filter parameters are not part of the QueryKey (except for the tenant identifier). So this test will fail:

```java
scope.inTransaction(
		session -> {
			session.enableFilter( "personName" ).setParameter( "name", "John" );
			Query<Person> query = session.createQuery(
					"from Person p",
					Person.class
			);
			query.setCacheable( true );

			List<Person> resultList = query.getResultList();
			assertThat( resultList ).hasSize( 1 );
			session.enableFilter( "personName" ).setParameter( "name", "John" );
		}
);
scope.inTransaction(
		session -> {
			session.enableFilter( "personName" ).setParameter( "name", "Jack" );
			Query<Person> query = session.createQuery(
					"from Person p",
					Person.class
			);
			query.setCacheable( true );

			List<Person> resultList = query.getResultList();
			assertThat( resultList ).hasSize( 0 ); //fails here with 1
		}
);
```

Should I open another bug for this?

Reply:
>What I'm trying to test is that QueryKey is properly serializable so it can be used as a cache key => if it's not then the query is not cached, there's a miss the second time, my test validates that
>I don't understand what your test is trying to validate: how is mutating the filter relevant in this case?

Your original fix was to create a copy of the `enabledFilterNames`, and while looking through the code, I saw that this is necessary because the underlying collection can be mutated. Even if `HashMap.KeySet` is not serializable, the problem is deeper than just the serializability concern.

>Should I open another bug for this?

IMO the `parameterBindingsMemento` should actually contain the filter parameter values. Yes, please open a new JIRA ticket for this.

Reply:
OK, I could finally reproduce the QueryKey mutability issue reliably, but it requires a second filter to be enabled after 
the query, otherwise an emptySet() is passed to the QueryKey the first time. So your test becomes this:
```java
scope.inTransaction(
		session -> {
			session.enableFilter( "personName" )
					.setParameter( "name", "John" );
			Query<Person> query = session.createQuery(
					"from Person p",
					Person.class
			);
			query.setCacheable( true );

			List<Person> resultList = query.getResultList();
			assertThat( resultList ).hasSize( 1 );

			session.enableFilter( "personName2" )
					.setParameter( "name", "John" );
		}
);
scope.inTransaction(
		session -> {
			session.enableFilter( "personName" )
					.setParameter( "name", "John" );
			Query<Person> query = session.createQuery(
					"from Person p",
					Person.class
			);
			query.setCacheable( true );

			List<Person> resultList = query.getResultList();
			assertThat( resultList ).hasSize( 1 );

			assertThat( getQueryCacheRegionStatistics( session ).getHitCount() ).isEqualTo( 1 );
		}
);
```

Also, the problem does not happen when using a jcache-compatible cache provider, as the QueryKey is serialized before comparison. So I've added the test in the hibernate-core module, that uses an in-memory cache backed by a simple ConcurrentMap

I still need to keep my test in the hibernate-cache module to validate the serialization issue.

I've also created HHH-16617 for the filter parameters missing from the cache key.